pkg/manager/utils/statefulset_builder.go

```diff
@@ -0,0 +1,139 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"github.com/pingcap/tidb-operator/pkg/util"
+
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type ContainerBuilder struct {
+	prototype *corev1.Container
+}
+
+func NewContainerBuilder(container *corev1.Container) *ContainerBuilder {
+	return &ContainerBuilder{
+		prototype: container,
+	}
+}
+
+func (cb *ContainerBuilder) Get() *corev1.Container {
+	return cb.prototype
+}
+
+func (cb *ContainerBuilder) Clone() *corev1.Container {
+	return cb.prototype.DeepCopy()
+}
+
+func (cb *ContainerBuilder) AddVolumeMounts(mounts ...corev1.VolumeMount) {
+	cb.prototype.VolumeMounts = append(cb.prototype.VolumeMounts, mounts...)
+}
+
+func (cb *ContainerBuilder) AddEnvs(envs ...corev1.EnvVar) {
+	cb.prototype.Env = util.AppendEnv(cb.prototype.Env, envs)
+}
+
+type PodTemplateSpecBuilder struct {
+	prototype *corev1.PodTemplateSpec
+}
+
+func NewPodTemplateSpecBuilder(podTemplate *corev1.PodTemplateSpec) *PodTemplateSpecBuilder {
+	return &PodTemplateSpecBuilder{
+		prototype: podTemplate,
+	}
+}
+
+func (pb *PodTemplateSpecBuilder) Get() *corev1.PodTemplateSpec {
+	return pb.prototype
+}
+
+func (pb *PodTemplateSpecBuilder) Clone() *corev1.PodTemplateSpec {
+	return pb.prototype.DeepCopy()
+}
+
+// ContainerBuilder return the specified container builder
+//
+// NOTE: don't save the builder, it will be outdated if container slice expand
+func (pb *PodTemplateSpecBuilder) ContainerBuilder(name string) *ContainerBuilder {
+	for i := range pb.prototype.Spec.Containers {
+		if pb.prototype.Spec.Containers[i].Name == name {
+			return NewContainerBuilder(&pb.prototype.Spec.Containers[i])
+		}
+	}
+	return nil
+}
+
+// AddContainers add container to pod's spec
+//
+// NOTE: if container slice expand, old ContainerBuilder reference will be outdated.
+func (pb *PodTemplateSpecBuilder) AddContainers(containers ...corev1.Container) {
+	pb.prototype.Spec.Containers = append(pb.prototype.Spec.Containers, containers...)
+}
+
+func (b *PodTemplateSpecBuilder) AddVolumes(volumes ...corev1.Volume) {
+	b.prototype.Spec.Volumes = append(b.prototype.Spec.Volumes, volumes...)
+}
+
+func (b *PodTemplateSpecBuilder) AddLabels(labels map[string]string) {
+	b.prototype.Labels = util.CombineStringMap(b.prototype.Labels, labels)
+}
+
+func (b *PodTemplateSpecBuilder) AddAnnotations(annos map[string]string) {
+	b.prototype.Annotations = util.CombineStringMap(b.prototype.Annotations, annos)
+}
+
+func (b *PodTemplateSpecBuilder) RunInHostNetwork() {
+	b.prototype.Spec.HostNetwork = true
+	b.prototype.Spec.DNSPolicy = corev1.DNSClusterFirstWithHostNet
+	for _, container := range b.prototype.Spec.Containers {
+		container.Env = util.AppendEnv(container.Env, []corev1.EnvVar{
+			{
+				Name: "POD_NAME",
```

Comment:
Why only add this env when using hostNetwork?
This is used as the advertise address, right?

Reply:
yes, advertise address contains pod name

Reply:
So why add the env here, it should be included by default.

Reply:
copy it to have same code with other managers, ngm manager will add this env in `GenerateStatefulSet`